### PR TITLE
Zachmu/funcs

### PR DESCRIPTION
**Pull request by zachmu**
Got rid of all embedded function fields in function types, since they make it impossible for analyzer to finish (function fields are not comparable with reflect.DeepEquals, which the analyzer uses to decide if the query plan has settled).

### CODE REVIEW

File: sql/expression/function/function.go [hunk 1]
```diff
@@ -38,43 +30,19 @@ func (uf *UnaryFunc) FunctionName() string {
 }
 
 // Eval implements the Expression interface.
```
Comment (andrew-wm-arthur) [inline sql/expression/function/function.go:32]: stale comment doc

File: enginetest/queries.go [hunk 2]
```diff
@@ -1299,6 +1299,103 @@ var QueryTests = []QueryTest{
...
+			{"1"},
+		},
+	},
+	{
+		Query:    "SELECT BIT_LENGTH(i) from mytable order by i limit 1",
+		Expected: []sql.Row{
+			{64},
+		},
+	},
+	// TODO: add additional tests for other functions
```
Comment (andrew-wm-arthur) [inline enginetest/queries.go:1398]: do we need a test for _every_ function? to ensure they can be analyzed?

File: sql/expression/function/math.go [hunk 3]
```diff
@@ -129,129 +127,361 @@ func (r *Rand) Eval(ctx *sql.Context, row sql.Row) (interface{}, error) {
...
-// RadiansFunc implements the radians function logic
-func RadiansFunc(_ *sql.Context, val interface{}) (interface{}, error) {
 	n, err := sql.Float64.Convert(val)
-
 	if err != nil {
 		return nil, err
 	}
 
-	return (n.(float64) * math.Pi) / 180.0, nil
+	return (n.(float64) * 180.0) / math.Pi, nil
```
Comment (andrew-wm-arthur) [inline sql/expression/function/math.go:416]: the diff was basically perfect up to here, good run.

File: sql/expression/function/math_test.go [hunk 4]
```diff
@@ -170,12 +170,12 @@ func TestCRC32(t *testing.T) {
 }
 
 func TestTrigFunctions(t *testing.T) {
-	asin := NewUnaryFunc("asin", sql.Float64, ASinFunc)
-	acos := NewUnaryFunc("acos", sql.Float64, ACosFunc)
-	atan := NewUnaryFunc("tan", sql.Float64, ATanFunc)
-	sin := NewUnaryFunc("sin", sql.Float64, SinFunc)
-	cos := NewUnaryFunc("cos", sql.Float64, CosFunc)
-	tan := NewUnaryFunc("atan", sql.Float64, TanFunc)
```
Comment (andrew-wm-arthur) [inline sql/expression/function/math_test.go:178]: bug fix here for correcting `NewUnaryFunc("tan", sql.Float64, ATanFunc)` and `NewUnaryFunc("atan", sql.Float64, TanFunc)`?

File: sql/expression/function/registry.go [hunk 5]
```diff
@@ -11,14 +11,14 @@ import (
...
 	sql.Function1{Name: "array_length", Fn: NewArrayLength},
-	NewUnaryFunc("ascii", sql.Uint8, AsciiFunc),
-	NewUnaryFunc("asin", sql.Float64, ASinFunc),
-	NewUnaryFunc("atan", sql.Float64, ATanFunc),
+	sql.Function1{Name: "ascii", Fn: NewAscii},
+	sql.Function1{Name: "asin", Fn: NewAsin},
+	sql.Function1{Name: "atan", Fn: NewAtan},
 	sql.Function1{Name: "avg", Fn: func(e sql.Expression) sql.Expression { return aggregation.NewAvg(e) }},
-	NewUnaryFunc("bin", sql.Text, BinFunc),
-	NewUnaryFunc("bit_length", sql.Int32, BinFunc),
```
Comment (andrew-wm-arthur) [inline sql/expression/function/registry.go:21]: bug fix for double `BinFunc`?

**Review by andrew-wm-arthur — APPROVED** [review]
LGTM! The nice diff made this pretty easy to review

File: sql/expression/function/registry.go (hunk 5 above)
Comment (zachmu) [reply sql/expression/function/registry.go:21]: yup

File: sql/expression/function/math_test.go (hunk 4 above)
Comment (zachmu) [reply sql/expression/function/math_test.go:178]: yup

File: enginetest/queries.go (hunk 2 above)
Comment (zachmu) [reply enginetest/queries.go:1398]: Yes, exactly. I'll clarify.

**Pull request merged**